distribution/client/src/lib/update.js

```diff
@@ -99,18 +100,89 @@ class Update {
+   * 1) restart Jenkins
+   * 2) Check health
+   * 3) if OK, continue
+   * 4) if KO, rollback state + go back to previous UL.
+   *
+   * But what if Jenkins does not restart even after rollback?
+     we will not try to go back further. But the client will stay
+
+     Open questions:
+     * how to avoid going again through this UL that borked the system,
+       while it's not been yet marked as tainted in the backend?
+     * how to report that borked case in a clear way
+  */
+  restartJenkins(rollingBack) { // Add param to stop recursion?
+    Supervisord.restartProcess('jenkins');
+
+    const options = {
+      timeoutInSeconds: 60
+    };
+    UI.publish('Jenkins should now be online, health checking');// (timeout=${options.timeoutInSeconds})`);
+
+    logger.info('Jenkins has restarted, now healthchecking!');
+
+    // FIXME: actually now I'm thinking throwing in HealthChecker might provide a more
+    // consistent promise usage UX here.
+    // checking healthState.health value is possibly a bit convoluted (?)
+    this.healthChecker.check(options)
+      .then( healthState => {
+        if (healthState.healthy) {
+          logger.info('Jenkins healthcheck after restart succeeded! Yey.');
+        } else {
+
+          if (rollingBack) {
+
+            // if things are wrong twice, stop trying and just holler for help
+            // Quick notice sketch, but I do think we need a very complete and informative message
+            const failedToRollbackMessage =
+              'Ooh noes :-(. We are terribly sorry but it looks like Jenkins failed to ' +
+              'upgrade, but even after the automated rollback we were unable to bring ' +
+              'to life. Please report this issue to the Jenkins Evergreen team. ' +
+              'Do not shutdown your instance as we have been notified of this failure ' +
+              'and are trying to understand what went wrong to push a new update that ' +
+              'will fix things.';
+            logger.error(failedToRollbackMessage);
+            UI.publish(failedToRollbackMessage);
+
+            // Not throwing an Error here as we want the client to keep running and ready
+            // since the next available UL _might_ fix the issue
+          } else {
+
+            const errorMessage = `Jenkins detected as unhealthy: ${healthState.message}. ` +
+                                 'Rolling back to previous update level.';
+            UI.publish(errorMessage);
+            logger.warn(errorMessage);
+
+            this.snapshotter.revertToLevelBefore(this.getCurrentLevel());
+            this.revertToPreviousUpdateLevel();
+            this.restartJenkins(true);
+          }
+        }
+        Storage.removeBootingFlag();
+      }).catch((errors) => {
+        logger.warn(`TODO ${errors}`);
+      }); // TODO catch?
+  }
+
+  revertToPreviousUpdateLevel() {
+    logger.error(`[NOT IMPLEMENTED YET] Revert UL-${this.getCurrentLevel()} to previous Update Level`);
```

Comment:
To be implemented. 

As discussed yesterday with @rtyler, we need to do the following:

* call the to be written new `this.taintCurrentLevel()` function
* then retrigger a `query()`, which should serve the *previous* update.